### PR TITLE
Support deep nested structures for mapMultiRowFields

### PR DESCRIPTION
**Pull request by mcosta91**
Hi! Thank you for you library.
I would like to improve a bit the current implementation of `mapMultiRowFields` to support deep nested fields.

A few things to note in this PR are:
 - I've added a single small dependency (traverse) to traverse the actual state object in order to create a `kind of` reactive structure. Maybe I can just import the code directly from it, if you think that's the right approach.
- I've abused a bit of closures to avoid unnecessary parsing.
- The path created for `updateField` and `getField` uses dot-notation even for arrays, so we have a strange `array.1.field`, but it still plays nicely with the current implementation of those getters and setters.

### CODE REVIEW

**Comment by maoberlehner** [comment]
Thank you for your contribution. I'll take a look at it!

File: package.json [hunk 1]
```diff
@@ -64,5 +64,8 @@
       "/node_modules/",
       "/test/package/"
     ]
+  },
+  "dependencies": {
+    "traverse": "^0.6.6"
```
Comment (maoberlehner) [inline package.json]: I'm not a huge fan of adding a dependency. Is this possible without (and without having to write a lot of custom code)?

File: src/index.js [hunk 2]
```diff
@@ -1,5 +1,52 @@
+import traverse from 'traverse';
 import arrayToObject from './lib/array-to-object';
 
+function computeTree(store, obj, mutationType, getterType, rootPath) {
+  // eslint-disable-next-line no-param-reassign
+  obj = JSON.parse(JSON.stringify(obj));
```
Comment (maoberlehner) [inline src/index.js]: Why is this necessary?

File: src/index.js [hunk 3]
```diff
@@ -63,26 +110,27 @@ export const mapMultiRowFields = normalizeNamespace((
...
+        return computedEntry;
+      },
+      set(value) {
+        const store = this.$store;
+        store.commit(mutationType, { path, value });
+
+        if (typeof value !== `object`) { store.commit(mutationType, { path, value }); } else {
+          const state = store.getters[getterType](path);
+          computedEntry = computeTree(store, state, mutationType, getterType, path);
+        }
```
Comment (maoberlehner) [inline src/index.js]: Please format `if` `else` like so:

```js
if (/*...*/) {
// ...
} else {
// ..
}
```

File: test/multi-row.test.js [hunk 4]
```diff
@@ -79,3 +79,77 @@ describe(`Component initialized with multi row setup.`, () => {
...
+
+  test(`It should update the store when the field values are updated. blau`, () => {
+    wrapper.find(`.street-0`).element.value = `New Street`;
+    wrapper.find(`.street-0`).trigger(`input`);
+    wrapper.find(`.number-1`).element.value = `43`;
+    wrapper.find(`.number-1`).trigger(`input`);
+    expect(store.state.users[0].addresses[0].street).toBe(`New Street`);
+    expect(store.state.users[1].addresses[0].number).toBe(`43`);
+  });
+});
```
Comment (maoberlehner) [inline test/multi-row.test.js]: Thank you for not forgetting about adding new Tests 👍

**Comment by mcosta91** [comment]
> I'm not quite sure if I understand it 100%
> Also I'm not sure if I want to add this because it adds a lot of complexity and as I said, I don't understand why this is needed.
> Please explain.
> 
> Thank you very much for your effort!

Oh sorry by not being more clear. What I'm proposing is supporting nested object structures for mapMultiRowFields, so we can have eg:

```html
        <div>
          <div v-for="(user, index) in users">
            <div v-for="address in user.addresses">
              <input :class="'street-' + index" v-model="address.street">
              <input :class="'number-' + index" v-model="address.number">
            </div>
          </div>
        </div>
```
or

```html
        <div>
          <div v-for="(user, index) in users">
            <input :class="'street-' + index" v-model="user.addresses[0].street">
            <input :class="'number-' + index" v-model="user.addresses[0].number">
          </div>
        </div>
```
being `users` a field mapped with mapMultiRowFields, so we can get, for exemple, https://github.com/maoberlehner/vuex-map-fields/issues/31#issue-331465650 working. 

My issue specifically is trying to get

```
<template>
    <div v-for="(formUnit, index) in formUnits" :key="index">
         <input type="text" class="listing-units__input" label="Valor" v-model="formUnit.pricingInfos[0].price"/>
     </div>
</template>
<script>
(...)
  computed: {
    ...mapMultiRowFields('listingForm', ['formUnits']),
  },
(...)
</script>
```

to work, but in the current implementation, changing input value doesn't trigger store mutation.

File: package.json (hunk 1 above)
Comment (mcosta91) [reply package.json]: ` Is this possible without (and without having to write a lot of custom code)?`
Sure. What we actually need is a pretty simple traverse algorithim. I can write and embed it if you agree with the general idea of this PR.

File: src/index.js (hunk 2 above)
Comment (mcosta91) [reply src/index.js]: It's just a `lazy` way of cloning data structures. We need it since we're changing the original data structure to accomodate getters/setters that trigger store mutations, while the original data structure is stored in vuex `state`. Another way of doing it is to make the traversal algorithm create new objects instead of mutating it, so we don't need to use cloning strategies.

**Comment by toddpress** [comment]
IMHO, @mcosta91's PR complete's this amazing tool. Just to speak to it's usefulness... My specific use-case is our store data consists partially of nested arrays of objects containing that contain object arrays-- it's hastily json-ified xml. In our case, unfortunately, we've no control over this admittedly janky format; however, @mcosta91's mapMRF works seamlessly with our tortuous data model! I'd argue it's worth the dependency in the case of traverse if this is a democracy. ;) Thanks to you both

**Comment by Tallyrald** [comment]
@maoberlehner Dear author, may I ask what is the status of this feature request? Is it denied? Is it considered?
I am working on a quite large project using a serious amount of complex data structures and so far the only issue I could not quite tackle is handling nested arrays of objects. Unfortunately changing the data structure is not possible.
I have tried Pathify, however it does not provide a solution for two-way binding in nested structures (in fact I do not know of any kind of solution at all that allows two-way v-model binding in nested structures like this), only reading from them and since I used vuex-map-field throughout the whole project so far, I am hesitant on pulling in another solution for this single use-case and would save me a lot of time if vuex-map-fields was capable of saving nested stuff.
Thank you for this amazing library!

**Comment by maoberlehner** [comment]
@Tallyrald I'm thinking about it. Not too happy with some of the details of the pull request. I add more comments to clarify.

File: package.json (hunk 1 above)
Comment (maoberlehner) [reply package.json]: Currently this package has no dependencies and I want to keep it that way. If you could write a very simple custom solution which does not add a lot of complexity, only for this specific use-case, it would be great!

File: src/index.js (hunk 2 above)
Comment (maoberlehner) [reply src/index.js]: I used this strategy myself in the past to clone objects and in certain situations I'm fine with it because sometimes we have to take shortcuts in our projects. But because this is a library used by hundreds of people, I do not feel comfortable with this "hack".

**Comment by wingsico** [comment]
> @maoberlehner Dear author, may I ask what is the status of this feature request? Is it denied? Is it considered?
> I am working on a quite large project using a serious amount of complex data structures and so far the only issue I could not quite tackle is handling nested arrays of objects. Unfortunately changing the data structure is not possible.
> I have tried Pathify, however it does not provide a solution for two-way binding in nested structures (in fact I do not know of any kind of solution at all that allows two-way v-model binding in nested structures like this), only reading from them and since I used vuex-map-field throughout the whole project so far, I am hesitant on pulling in another solution for this single use-case and would save me a lot of time if vuex-map-fields was capable of saving nested stuff.
> Thank you for this amazing library!

hello, I found some problem of this feature request: if you try to use it in nested structure, it works. but if you do it with one layer array, it's not reactive, I have uploaded a npm package of this pr named '@talkc/vuex-map-fields', you can try it and you will find the bug too

**Comment by Xationch** [comment]
What's the status of deep nested ? I really need this feature.

**Comment by lhy666** [comment]
I really need this feature. too
